src/sources/file_descriptors/mod.rs

```diff
@@ -0,0 +1,138 @@
+use std::{io, thread};
+
+use async_stream::stream;
+use bytes::Bytes;
+use chrono::Utc;
+use codecs::{
+    decoding::{DeserializerConfig, FramingConfig},
+    StreamDecodingError,
+};
+use futures::{channel::mpsc, executor, SinkExt, StreamExt};
+use tokio_util::{codec::FramedRead, io::StreamReader};
+use vector_core::config::LogNamespace;
+use vector_core::ByteSizeOf;
+
+use crate::{
+    codecs::DecodingConfig,
+    config::log_schema,
+    internal_events::{BytesReceived, OldEventsReceived, StreamClosedError},
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+
+#[cfg(all(unix, feature = "sources-pipe"))]
+pub mod pipe;
+#[cfg(feature = "sources-stdin")]
+pub mod stdin;
+
+pub trait FileDescriptorConfig {
+    fn host_key(&self) -> Option<String>;
+    fn framing(&self) -> Option<FramingConfig>;
+    fn decoding(&self) -> DeserializerConfig;
+    fn description(&self) -> String;
+}
+
+pub fn file_descriptor_source<R, C>(
+    mut reader: R,
+    config: C,
+    shutdown: ShutdownSignal,
+    mut out: SourceSender,
+    name: &'static str,
+) -> crate::Result<crate::sources::Source>
+where
+    R: Send + io::BufRead + 'static,
+    C: FileDescriptorConfig,
+{
+    let host_key = config
+        .host_key()
+        .unwrap_or_else(|| log_schema().host_key().to_string());
+    let hostname = crate::get_hostname().ok();
+
+    let decoding = config.decoding();
+    let framing = config
+        .framing()
+        .unwrap_or_else(|| decoding.default_stream_framing());
+    let decoder = DecodingConfig::new(framing, decoding, LogNamespace::Legacy).build();
+
+    let (mut sender, receiver) = mpsc::channel(1024);
+
+    // Spawn background thread with blocking I/O to process fd.
+    //
+    // This is recommended by Tokio, as otherwise the process will not shut down
+    // until another newline is entered. See
+    // https://github.com/tokio-rs/tokio/blob/a73428252b08bf1436f12e76287acbc4600ca0e5/tokio/src/io/stdin.rs#L33-L42
+    let description = config.description();
+    thread::spawn(move || {
+        info!("Capturing {}.", description);
+
+        loop {
+            let (buffer, len) = match reader.fill_buf() {
+                Ok(buffer) if buffer.is_empty() => break, // EOF.
+                Ok(buffer) => (Ok(Bytes::copy_from_slice(buffer)), buffer.len()),
+                Err(error) if error.kind() == std::io::ErrorKind::Interrupted => continue,
+                Err(error) => (Err(error), 0),
+            };
+
+            reader.consume(len);
+
+            if executor::block_on(sender.send(buffer)).is_err() {
+                // Receiver has closed so we should shutdown.
+                break;
+            }
+        }
```

Comment:
Could these inner closures be broken out into separate functions to improve the readability of this function? Particularly the one below.

Reply:
@bruceg you are asking to extract the contents of

```Rust
thread::spawn(move || {
  // <this>
})
```

and

```Rust
Ok(Box::pin(async move {
   // <this>
})
```

, into helper functions correct ?

Reply:
Yes, if it makes sense. Usually it's fairly straightforward (everything that is `move`d becomes a parameter) but occasionally the amount of captured variables becomes unwieldy.